plugins/hls-splice-plugin/src/Ide/Plugin/Splice.hs

```diff
@@ -0,0 +1,271 @@
+                Left err ->
+                    (Left $ responseError $ T.pack err, Nothing)
+                Right edits ->
+                    ( Right Null
+                    , Just (WorkspaceApplyEdit, ApplyWorkspaceEditParams edits)
+                    )
+    where
+        defaultResult = (Right Null, Nothing)
+
+findSuperSpansAsc :: SrcSpan -> [(LHsExpr GhcTc, a)] -> [(SrcSpan, a)]
+findSuperSpansAsc srcSpan =
+    sortOn (SubSpan . fst)
+        . mapMaybe
+            ( \(L spn _, e) -> do
+                guard (srcSpan `isSubspanOf` spn)
+                pure (spn, e)
+            )
+
+class (Outputable (ast GhcRn), ASTElement (ast GhcPs)) => HasSplice ast where
+    type SpliceOf ast :: Kinds.Type -> Kinds.Type
+    type SpliceOf ast = HsSplice
+    matchSplice :: Proxy# ast -> ast GhcPs -> Maybe (SpliceOf ast GhcPs)
+
+instance HasSplice HsExpr where
+    matchSplice _ (HsSpliceE _ spl) = Just spl
+    matchSplice _ _ = Nothing
+
+instance HasSplice Pat where
+    matchSplice _ (SplicePat _ spl) = Just spl
+    matchSplice _ _ = Nothing
+
+instance HasSplice HsType where
+    matchSplice _ (HsSpliceTy _ spl) = Just spl
+    matchSplice _ _ = Nothing
+
+-- | FIXME:  Is thereAny "clever" way to do this exploiting TTG?
+reportEditor :: MonadIO m => LspFuncs a -> MessageType -> [T.Text] -> m ()
+reportEditor lsp msgTy msgs =
+    liftIO $
+        sendFunc lsp $
+            NotShowMessage $
+                NotificationMessage "2.0" WindowShowMessage $
+                    ShowMessageParams msgTy $
+                        T.unlines msgs
+
+unRenamedE ::
+    forall ast m.
+    (Fail.MonadFail m, HasSplice ast) =>
+    DynFlags ->
+    ast GhcRn ->
+    TransformT m (Located (ast GhcPs))
+unRenamedE dflags expr = do
+    uniq <- show <$> uniqueSrcSpanT
+    (anns, expr') <-
+        either (fail . show) pure $
+            parseAST @(ast GhcPs) dflags uniq $
+                showSDoc dflags $ ppr expr
+    let _anns' = setPrecedingLines expr' 0 1 anns
+    -- modifyAnnsT $ mappend anns'
+    pure expr'
+
+-- TODO: workaround when HieAst unavailable (e.g. when the module itself errors)
+-- TODO: Declaration Splices won't appear in HieAst; perhaps we must just use Parsed/Renamed ASTs?
+codeAction :: CodeActionProvider IdeState
+codeAction _ state plId docId range0 _ =
+    fmap (maybe (Right $ List []) Right) $
+        runMaybeT $ do
+            fp <- MaybeT $ pure $ uriToNormalizedFilePath $ toNormalizedUri theUri
+            (getAsts . hieAst -> asts, posMap) <-
+                MaybeT . runAction "splice.codeAction.GitHieAst" state $
+                    useWithStale GetHieAst fp
+            ran' <-
+                MaybeT $
+                    pure $
+                        fromCurrentRange posMap range0
+            fmap (List . DL.toList) $
+                execWriterT $
+                    getAp $ ifoldMap (fmap Ap . go ran') asts
+    where
+        theUri = docId ^. J.uri
+        go ran' fs ast =
```

Comment:
I think you don't need the `HieAST` now, you can just look at the source spans of the original unexpanded splices.

Reply:
Done! There is a pitfall that HieAST and ParsedModule gives slightly different Locations for splices, resulting in the change of searching strategy for splice span. As an additional benefit, we can now invoke the CodeAction over whole splices; formerly we can invoke CodeLens only strictly inside splices (e.g. right after `(` from before `)` in `$(foo)`).